Add golangci-lint to list of tools

Jasstkn opened the pull request:
Signed-off-by: Maria Kotlyarevskaya <mariia.kotliarevskaia@gmail.com>

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

Fixes #730 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [x] I have raised an issue to propose this change, which has been given a label of `design/approved` by a maintainer ([required](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md))


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

```bash
$ ./arkade get golangci-lint
Downloading: golangci-lint
2022/08/11 12:34:22 Looking up version for golangci-lint
2022/08/11 12:34:22 Found: v1.48.0
Downloading: https://github.com/golangci/golangci-lint/releases/download/v1.48.0/golangci-lint-1.48.0-darwin-arm64.tar.gz
9.14 MiB / 9.14 MiB [----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00%
/var/folders/t7/dp293qn57h12csyb_b4zygtw0000gp/T/golangci-lint-1.48.0-darwin-arm64.tar.gz written.
2022/08/11 12:34:24 Looking up version for golangci-lint
2022/08/11 12:34:24 Found: v1.48.0
2022/08/11 12:34:25 Extracted: /var/folders/t7/dp293qn57h12csyb_b4zygtw0000gp/T/golangci-lint
2022/08/11 12:34:25 Copying /var/folders/t7/dp293qn57h12csyb_b4zygtw0000gp/T/golangci-lint to /Users/maria.kotlyarevskaya/.arkade/bin/golangci-lint

Wrote: /Users/maria.kotlyarevskaya/.arkade/bin/golangci-lint (25.87MB)

# Add arkade binary directory to your PATH variable
export PATH=$PATH:$HOME/.arkade/bin/

# Test the binary:
/Users/maria.kotlyarevskaya/.arkade/bin/golangci-lint

# Or install with:
sudo mv /Users/maria.kotlyarevskaya/.arkade/bin/golangci-lint /usr/local/bin/

🐳 arkade needs your support, learn more: https://my.arkade.dev

$ ./arkade get golangci-lint --arch amd64 --os darwin                                                                                                                                                                     
Downloading: golangci-lint
2022/08/11 12:56:38 Looking up version for golangci-lint
2022/08/11 12:56:39 Found: v1.48.0
Downloading: https://github.com/golangci/golangci-lint/releases/download/v1.48.0/golangci-lint-1.48.0-darwin-amd64.tar.gz
9.40 MiB / 9.40 MiB [----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00%
/var/folders/t7/dp293qn57h12csyb_b4zygtw0000gp/T/golangci-lint-1.48.0-darwin-amd64.tar.gz written.
2022/08/11 12:56:40 Looking up version for golangci-lint
2022/08/11 12:56:40 Found: v1.48.0
2022/08/11 12:56:41 Extracted: /var/folders/t7/dp293qn57h12csyb_b4zygtw0000gp/T/golangci-lint
2022/08/11 12:56:41 Copying /var/folders/t7/dp293qn57h12csyb_b4zygtw0000gp/T/golangci-lint to /Users/maria.kotlyarevskaya/.arkade/bin/golangci-lint

Wrote: /Users/maria.kotlyarevskaya/.arkade/bin/golangci-lint (25.94MB)

# Add arkade binary directory to your PATH variable
export PATH=$PATH:$HOME/.arkade/bin/

# Test the binary:
/Users/maria.kotlyarevskaya/.arkade/bin/golangci-lint

# Or install with:
sudo mv /Users/maria.kotlyarevskaya/.arkade/bin/golangci-lint /usr/local/bin/

🐳 arkade needs your support, learn more: https://my.arkade.dev

# test tool
$ /Users/maria.kotlyarevskaya/.arkade/bin/golangci-lint --version

golangci-lint has version 1.48.0 built from 2d8fea81 on 2022-08-04T18:44:38Z
Smart, fast linters runner. Run it in cloud for every GitHub pull request on https://golangci.com

Usage:
  golangci-lint [flags]
  golangci-lint [command]

Available Commands:
  cache       Cache control and information
  completion  Generate the autocompletion script for the specified shell
  config      Config
  help        Help
  linters     List current linters configuration
  run         Run the linters
  version     Version

Flags:
      --color string              Use color when printing; can be 'always', 'auto', or 'never' (default "auto")
  -j, --concurrency int           Concurrency (default NumCPU) (default 8)
      --cpu-profile-path string   Path to CPU profile output file
  -h, --help                      help for golangci-lint
      --mem-profile-path string   Path to memory profile output file
      --trace-path string         Path to trace output file
  -v, --verbose                   verbose output
      --version                   Print version

Use "golangci-lint [command] --help" for more information about a command.
```

## Are you a GitHub Sponsor yet (Yes/No?)

<!-- Requests from sponsors take priority -->
<!--- Check at https://github.com/sponsors/alexellis         -->

- [ ] Yes
- [x] No

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Documentation

- [x] I have updated the list of tools in README.md if (required) with `./arkade get -o markdown`
- [ ] I have updated the list of apps in README.md if (required) with `./arkade install --help`

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--- In case it is a new application -->
- [x] I have tested this on arm, or have added code to prevent deployment

Shikachuu commented on pkg/get/tools.go:
```diff
@@ -2545,5 +2545,29 @@ https://github.com/{{.Owner}}/{{.Repo}}/releases/download/{{.Version}}/{{.Name}}
+							{{$osStr = "darwin"}}
+							{{- end -}}
+							{{$archStr := .Arch}}
+							{{- if eq .Arch "aarch64" -}}
```
In case of a `darwin`, this isn't `aarch64` but `arm64` can you add that as an `or` to to your condition please?

Jasstkn replied:
Added and clean up naming a bit.

Jasstkn commented:
Hey. I have a question as I'm not sure what is the right thing to do: golangci-lint provides a number of different binaries. Should I make them supported in the arkade? I didn't find any information about supported architecture/os except of [here](https://github.com/alexellis/arkade#download-cli-tools-with-arkade).

alexellis commented:
Hi @Jasstkn - nice PR and thank you for taking the time to read and then fill out the PR template.

I would say we need armv7 also, and you're right that we don't say which binaries are expected in our contribution guide, yet. But we will fix that thanks to your suggestion.

[golangci-lint-1.48.0-linux-armv7.tar.gz](https://github.com/golangci/golangci-lint/releases/download/v1.48.0/golangci-lint-1.48.0-linux-armv7.tar.gz)

Alex

Shikachuu commented on pkg/get/get_test.go:
```diff
@@ -4359,3 +4359,56 @@ func Test_DownloadTerrascan(t *testing.T) {
+		},
+		{
+			os:      "darwin",
+			arch:    archARM64,
```
In case of synthetic tests the described `arm64` behaviour doesn't occur, to mimic a real world scenario, change the unit test's arch to the following const: `archDarwinARM64`.

alexellis replied:
@Jasstkn did you copy this from another example? If you did we may need to fix that one.

What @Shikachuu is saying is that we need the text `arm64` here which is in the `archDarwinARM64` const. MacOS decided to do things differently to Linux which uses the text "aarch64" in the `archARM64 const.

Jasstkn replied:
Yep, I looked up other test for similar change. I'll fix it right away!

Jasstkn replied:
fixed

Shikachuu reviewed (changes requested):
Sorry, but commit sign-off you must use your real name as described here: https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md#sign-off-your-work

Also, please check the other comments about how `darwin` and `arm64` works. 

All-in-all nice work!

alexellis commented:
I'm seeing a sign-off message on the commit. Are you?

[I added these guidelines](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md#recommended-architectures-for-binaries-in-arkade-get), feel free to add to them @Shikachuu

Shikachuu commented:
> I'm seeing a sign-off message on the commit. Are you?
> 
> [I added these guidelines](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md#recommended-architectures-for-binaries-in-arkade-get), feel free to add to them @Shikachuu

Yeah, there is a sign-off, but with a nickname instead of a real one.

It is a good idea to have those guildlines to avoid confusion, thanks I'll update them soon!

Jasstkn commented:
> > I'm seeing a sign-off message on the commit. Are you?
> > [I added these guidelines](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md#recommended-architectures-for-binaries-in-arkade-get), feel free to add to them @Shikachuu
> 
> Yeah, there is a sign-off, but with a nickname instead of a real one.
> 
> It is a good idea to have those guildlines to avoid confusion, thanks I'll update them soon!

fixed signing off commits.

Shikachuu reviewed (approved):
Last but not least, most of the time we like to keep one commit per PR or if you have more you should have a really good reason for it.
(Thats why we like to amend to a commit and commonly require rebase a multiple commit PR, before we approve/merge it.)

Please do a rebase, besides that LGTM!

Jasstkn commented:
> Last but not least, most of the time we like to keep one commit per PR or if you have more you should have a really good reason for it. (Thats why we like to amend to a commit and commonly require rebase a multiple commit PR, before we approve/merge it.)
> 
> Please do a rebase, besides that LGTM!

One commit it is then :)

Shikachuu commented on pkg/get/tools.go:
```diff
@@ -2545,5 +2545,31 @@ https://github.com/{{.Owner}}/{{.Repo}}/releases/download/{{.Version}}/{{.Name}}
+							{{- else if eq .Arch "armv7l" -}}
+							{{$arch = "armv7"}}
+							{{- end -}}
+							{{.Name}}-{{slice .Version 1}}-{{$os}}-{{$arch}}.tar.gz`,
```
Good catch @alexellis, I some how overlooked it.
It is used to trim the v from the version string right @Jasstkn?
For that you should use the `.VersionNumber` variable which only contains the number and drops the `v` char from the begining.

Jasstkn replied:
Replaced it with `.VersionNumber`, thanks

alexellis replied:
I thought that was what was happening here and glad we could switch to `.VersionNumber`

Something else to document in the contributing guide if you'd like to send a PR @Shikachuu ?

Shikachuu replied:
Most definetly!

alexellis reviewed (approved):
Approved

alexellis commented:
Thanks for the contribution @Jasstkn. Now that you're familiar with how the apps work, feel free to have a browse at other issues or to suggest new apps for K8s or CLIs that you think you could use?
